command/debug.go

```diff
@@ -0,0 +1,945 @@
+		resp, err := client.Logical().Read("sys/config/state/sanitized")
+		if err != nil {
+			captErr := newCaptureError("config", err)
+			c.debugIndex.Errors = append(c.debugIndex.Errors, captErr)
+
+			c.logger.Error("config: error capturing config state", "error", err)
+		}
+
+		if resp != nil && resp.Data != nil {
+			collection := []map[string]interface{}{
+				{
+					"timestamp": time.Now().UTC(),
+					"config":    resp.Data,
+				},
+			}
+			if err := c.persistCollection(collection, "config.json"); err != nil {
+				c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "config.json", err))
+			}
+		}
+	}
+
+	return nil
+}
+
+// capturePollingTargets captures all dynamic targets over the specified
+// duration and interval.
+func (c *DebugCommand) capturePollingTargets(client *api.Client) error {
+	startTime := time.Now()
+	durationCh := time.After(c.flagDuration + debugDurationGrace)
+
+	// Track current interval count to show progress.
+	// totalCount := int(c.flagDuration.Seconds()/c.flagInterval.Seconds()) + 1
+	var idxCount, mIdxCount int
+
+	// errCh is used to capture any polling errors and recorded in the index
+	errCh := make(chan *captureError)
+
+	var serverStatusCollection []map[string]interface{}
+	var replicationStatusCollection []map[string]interface{}
+	var metricsCollection []map[string]interface{}
+	var hostInfoCollection []map[string]interface{}
+
+	// Start a goroutine to collect data as soon as it's received
+	collectCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case metricsEntry := <-c.metricsCh:
+				metricsCollection = append(metricsCollection, metricsEntry)
+			case serverStatusEntry := <-c.serverStatusCh:
+				serverStatusCollection = append(serverStatusCollection, serverStatusEntry)
+			case replicationStatusEntry := <-c.replicationStatusCh:
+				replicationStatusCollection = append(replicationStatusCollection, replicationStatusEntry)
+			case hostInfoEntry := <-c.hostInfoCh:
+				hostInfoCollection = append(hostInfoCollection, hostInfoEntry)
+			case <-c.doneCh:
+				close(collectCh)
+				return
+			}
+		}
+	}()
+
+	// Start capture by capturing the first interval before we hit the first
+	// ticker.
+	c.pollingWg.Add(1)
+	go c.intervalCapture(client, idxCount, startTime, errCh)
+
+	c.pollingWg.Add(1)
+	go c.metricsIntervalCapture(client, mIdxCount, errCh)
+
+	// Capture at each interval, until end of duration or interrupt.
+	intervalTicker := time.Tick(c.flagInterval)
+	metricsIntervalTicker := time.Tick(c.flagMetricsInterval)
+POLLING:
+	for {
+		select {
+		case err := <-errCh:
+			c.debugIndex.Errors = append(c.debugIndex.Errors, err)
+		case <-intervalTicker:
+			idxCount++
+			c.pollingWg.Add(1)
+			go c.intervalCapture(client, idxCount, startTime, errCh)
+		case <-metricsIntervalTicker:
+			mIdxCount++
+			c.pollingWg.Add(1)
+			go c.metricsIntervalCapture(client, mIdxCount, errCh)
+		case <-durationCh:
+			// Wait for polling requests to finish before breaking
+			c.pollingWg.Wait()
```

Comment:
This seems liable to deadlock.  We're waiting for the goroutines to exit and decrement the waitgroup, but they may be trying to write to errCh, which is unbuffered and only consumed above (and can't be while we're waiting.)

Why even bother with errCh?  Why not just use a mutex and have the goroutines append errors to c.debugIndex.Errors themselves?

Same goes with the channels we harvest the data from, only there we wouldn't even need a mutex I think.

I also think things would be simplified if you moved the tickers into the goroutines and made them long-running instead of starting a new one for each tick.  Right now there's the possibility of having multiple captures overlap if the interval is set low enough relative to the time they take. 

To me this looks like a great candidate for `oklog/run.Group`.

Reply:
There could be a re-write to use `oklog/run.Group`, but due to time constraint I'd like to propose simply moving `c.pollingWg.Wait()` out of the switch case.

> Why even bother with errCh? Why not just use a mutex and have the goroutines append errors to c.debugIndex.Errors themselves?

Using a mutex would mean that there could be potential for race when logging the error long after the error has occurred if the lock is held long by some other goroutine, especially if the lock is no just used for error logging.

> Same goes with the channels we harvest the data from, only there we wouldn't even need a mutex I think.

You would need some sort of locking since it'd be writing to the same collection by different goroutines.

Reply:
> There could be a re-write to use oklog/run.Group, but due to time constraint I'd like to propose simply moving c.pollingWg.Wait() out of the switch case.

I'm mostly concerned with fixing the deadlock, but I saw an opportunity to simplify the design at the same time.  You may find it saves development time to do that even if it's a more invasive change, as this may not be the only problem and reasoning about interactions with a bunch of goroutines, channels, and waitgroups isn't easy.  Your call.

> Using a mutex would mean that there could be potential for race when logging the error long after the error has occurred if the lock is held long by some other goroutine, especially if the lock is no just used for error logging.

I wasn't suggesting using it for anything other than error logging.  I prefer fine-grained locks short-lived locks as they're much easier to reason about.  Alternatively you could have an err slice per collection method and no locking.

> You would need some sort of locking since it'd be writing to the same collection by different goroutines.

Who would be writing to the same collection?  I'm suggesting you make e.g. the slice `hostInfoCollection` a member of `DebugCommand`, then your code that collects host info would append to that instead of writing to `hostInfoCh`.

Reply:
> Who would be writing to the same collection? I'm suggesting you make e.g. the slice hostInfoCollection a member of DebugCommand, then your code that collects host info would append to that instead of writing to hostInfoCh.

The write to `hostInfoCollection`, which could theoretically mean that two different intervalCapture goroutines (or more precisely, the goroutines that it spawns) are writing to the same slice. The initial implementation was writing to a shared slice but the data race was caught by the race detector.

Reply:
I was assuming you were going to fix it so that couldn't happen.  There should not be two goroutines collecting the same data at the same time.  That's just a recipe for confusion.